Add Mercurial support

pepijnve opened the pull request:
At the moment this PR represents an initial implementation of Mercurial support.
I've created this already to get early feedback. Does this approach look like the proper way to integrate with Mercurial?

pepijnve committed:
Add Mercurial support

adrienthebo commented on lib/r10k/source/hg.rb:
```diff
@@ -0,0 +1,124 @@
+# repository and enumerating the branches for the Git repository. Branches
+# are mapped to environments without modification.
+#
+# @since 1.3.0
```
This version information is incorrect.

adrienthebo commented on lib/r10k/source/hg.rb:
```diff
@@ -0,0 +1,124 @@
+require 'r10k/environment/hg'
+require 'r10k/environment/name'
+
+# This class implements a source for Git environments.
```
Bit of a nitpick, but this comment will need to be updated.

adrienthebo commented on lib/r10k/source/hg.rb:
```diff
@@ -0,0 +1,124 @@
+
+  # @!attribute [r] type
+  #   @return [:hg] The type of this control repo source
+  attr_reader :type
```
Where is this used?

pepijnve replied:
I had made these changes on the r10k-api branch by accident at first. I rebased to master and this bit got carried over. I'll get remove it since it's not relevant on the master branch yet.

pepijnve committed:
Clean up Mercurial source implementation

adrienthebo commented on lib/r10k/hg/stateful_repository.rb:
```diff
@@ -0,0 +1,86 @@
+    case status
+      when :absent
+        logger.debug { "Cloning #{@repo.path} and checking out #{@rev}" }
+        @repo.clone(@cache.hg_dir.to_s, {:rev => sha})
```
The git implementation does a lot of work to avoid copying extra objects from the cached repo around; how efficient is cloning a local repo via hg?

pepijnve replied:
Unfortunately Mercurial doesn't allow the same level of sharing between clones that Git does. Local clones use hard links, but that's only for the initial clone. Pulling changes triggers copy-on-write logic that clears the hard links for any changed files. I don't think there's much that can be done about this.

I kept the cached local repo to avoid hitting the network for each environment.

pepijnve replied:
I was digging through the Mercurial man pages just now looking for a way to improve this. One easy improvement is passing the `--branch` parameter when calling `clone` and `pull` for the environment clones. That way we only clone/pull the data that is relevant for each environment rather than cloning everything for each environment.

pepijnve committed:
Use cachedir from config as cache_root for Mercurial

pepijnve committed:
Clone and push only the branch corresponding to an environment

pepijnve committed:
Add Mercurial module implementation

pepijnve committed:
Add support for Mercurial bookmarks

pepijnve committed:
Add initial Mercurial documentation

pepijnve committed:
Corrections to bookmark support

pepijnve committed:
Add Merurial dynamic environments documentation

pepijnve committed:
Make choice between branches and bookmarks mutually exclusive
While feasible to combine both, bookmarks often mean that named branches have multiple heads which would require some kind of heuristic to determine which branch head should be used. To keep things simple to explain the two models are mutually exclusive for the time being.

pepijnve committed:
Correct title in hg-environments.mkd
